pkg/dbconn/conn_test.go

```diff
@@ -59,3 +59,27 @@ func TestNewConn(t *testing.T) {
 	assert.Error(t, err)
 	assert.Nil(t, db)
 }
+
+func TestNewConnRejectsReadOnlyConnections(t *testing.T) {
+	testutils.RunSQL(t, "DROP TABLE IF EXISTS t1")
+	testutils.RunSQL(t, "CREATE TABLE t1 (a INT NOT NULL, b INT, c INT, PRIMARY KEY (a))")
+
+	db, err := New(testutils.DSN(), NewDBConfig())
+	assert.NoError(t, err)
+	defer db.Close()
+	_, err = db.Exec("set session transaction read only")
+	assert.NoError(t, err)
+
+	// This would error, but `database/sql` automatically retries on a
+	// new connection (because we enforce `rejectReadOnly=true`) which is not read-only, and eventually succeed.
+	// See also:
+	// docs: https://github.com/go-sql-driver/mysql?tab=readme-ov-file#rejectreadonly
+	// rejectReadOnly test in `go-sql-driver/mysql`: https://github.com/go-sql-driver/mysql/blob/52c1917d99904701db2b0e4f14baffa948009cd7/driver_test.go#L2270-L2301
+	_, err = db.Exec("insert into t1 values (1, 2, 3)")
+	assert.NoError(t, err)
+
+	var count int
+	err = db.QueryRow("select count(*) from t1 where a=1").Scan(&count)
+	assert.NoError(t, err)
+	assert.Equal(t, 1, count)
```

Comment:
Nice idea for a test. You could try:
- Setting the pool size to 1
- Setting a session variable in the connection as well
 
That way when the insert fails and a new connection is created, you can ensure the session variable is no longer set and that the connection is truly recycled.